### PR TITLE
PackratParsers#phrase: explicit return type (fixes #369)

### LINKED ISSUE #369: PackratParsers#phrase - different behaviour in Scala 2 vs Scala 3
- Repro/Test: https://github.com/ilya-klyuchnikov/scala-parser-combinators/commit/92a338ea19567614d30b1f6f1b8b7953849779a5
- It's OK for Scala 2.x, but fails for Scala 3.x - https://github.com/ilya-klyuchnikov/scala-parser-combinators/runs/2292531311

### PR DESCRIPTION
**Pull request by ilya-klyuchnikov**
This fixes #369.

*TL;DR* A missed result type for `PackratParsers#phrase` results in a different behaviour (Scala 2.x. vs Scala 3.x) for edgy cases.

Long explanation:

It seems (I was not able to quickly find a corresponding reference) that in Scala 3.0 if an overriding method doesn't explicitly specify the result type, the result type of the base method is taken (even if the compiler can figure out a more concrete type).

This leads to a corner case with PackratParsers#phrase, which in 2.x has the effective/inferred type `def phrase[T](p: Parser[T]): PackratParser[T]` and in 3.x has the effective/inferred type `def phrase[T](p: Parser[T]): Parser[T]`.

This results into an extra implicit conversion `parser2packrat` being inserted even if the current parser is already a Packrat Parser. This, in turn, breaks assumptions of the `memo` method.

---

Providing explicit return type fixes #369 subtlety.
- A corresponding test case was added.

---

The added test is OK for Scala 2.x, but fails for Scala 3.x - as can be seen from this run - https://github.com/ilya-klyuchnikov/scala-parser-combinators/commit/92a338ea19567614d30b1f6f1b8b7953849779a5.
Correspondingly, after the fix, the test is OK for both Scala 2.x and Scala 3.x

---
I have been bitten by this subtlety - while migrating https://github.com/ilya-klyuchnikov/tapl-scala to Scala 3.

### CODE REVIEW

**Comment by Philippus** [comment]
Nice find, @SethTisue is this something that may be of interest to the scala 3 team?

**Review by Philippus — APPROVED** [review]

**Comment by SethTisue** [comment]
the Scala 2 behavior is usually considered undesirable here, and I believe the Scala 3 behavior is intentional: https://github.com/scala/bug/issues/7212

I've opened a ticket suggesting the Scala 3 migration guide mention this: https://github.com/scalacenter/scala-3-migration-guide/issues/181

**Comment by SethTisue** [comment]
the Scala 2 behavior is usually considered undesirable here, and I believe the Scala 3 behavior is intentional: https://github.com/scala/bug/issues/7212

**Comment by SethTisue** [comment]
thank you Ilya!

**Comment by SethTisue** [comment]
Scala 3.0.0-RC3 is around the corner, so we'll publish the fix then.
